### DISCUSSION TITLE
WebSocket connection to 'wss://toot.igniterealtime.org/api/v1/streaming/?' failed

### DISCUSSION THREAD

**Discussion by ClaudeStabile** [mastodon/mastodon#23643] [Q&A]
### Steps to reproduce the problem

Hello,

To reproduce the issue, you need to create 2 mastodon docker instances running within same docker server.
Instances are started from different folders using a different docker-compose.xml.
The instance with almost all defaults naming and settings is working fine. Additionnal instances created does not work for wss streaming and show permanent errs in Java Console.

How can i fix this connectivity issue ? I provide attached docker-compose.xml + nginx config file that is failing

Many thanks in advance for providing any hints/help

### Expected behaviour

No wss connection errors in console and working

### Actual behaviour

Failed to retrieve streaming via wss for site https://toot.igniterealtime.org Wss Errs showing in console

### Detailed description

I have wss connectivity problem from 2 docker instances(mastodon.noalyss.eu and  toot.igniterealtime.org. I run several mastodon docker instances, the one with default settings is working fine but extra instances i try to deploy works except  for wss streamer connections.
Apparently it does not catch up proper credentials in extra instances for some reasons.

I did check within my docker containers DB_PASS DB_NAME, credentials are correctly sourced from database.env and application.env

Error showing up in console :
...
![screenshot47](https://user-images.githubusercontent.com/29485972/205015524-0b6948e0-d244-4e46-b176-44058146826d.png)

==> problems looks similar to https://github.com/mastodon/mastodon/issues/7553

# Any clue or hint to tackle issue would greatly help 

## More details : 
I run 3 docker instances 
![screenshot48](https://user-images.githubusercontent.com/29485972/205017678-42846dcd-545f-4c04-b3f3-330b3c445c19.png)

**My docker-compose.yml file** for igniterealtime

cat docker-compose.yml
version: '3'

services:
  postgresql:
    image: postgres:14-alpine 
    env_file: database.env
    restart: always
    shm_size: 256mb
    healthcheck:
      test: ['CMD', 'pg_isready', '-U', 'postgres']
    volumes:
      - postgresql:/var/lib/postgresql/data
    networks:
      - igniterealtime_internal_network

#  pgbouncer:
#    image: edoburu/pgbouncer:1.12.0
#    env_file: database.env
#    depends_on:
#      - postgresql
#    healthcheck:
#      test: ['CMD', 'pg_isready', '-h', 'localhost']
#    networks:
#      - igniterealtime_internal_network

  redis:
    image: redis:7-alpine 
    restart: always
    healthcheck:
      test: ['CMD', 'redis-cli', 'ping']
    volumes:
      - redis:/data
    networks:
      - igniterealtime_internal_network

  redis-volatile:
    image: redis:7-alpine 
    restart: always
    healthcheck:
      test: ['CMD', 'redis-cli', 'ping']
    networks:
      - igniterealtime_internal_network

  elasticsearch:
    image: elasticsearch:7.17.3
    restart: always
    env_file: database.env
    environment:
      - cluster.name=elasticsearch-mastodon
      - discovery.type=single-node
      - bootstrap.memory_lock=true
      - xpack.security.enabled=true
      - ingest.geoip.downloader.enabled=false
    ulimits:
      memlock:
        soft: -1
        hard: -1
    healthcheck:
      test: ["CMD-SHELL", "nc -z elasticsearch 9200"]
    volumes:
      - elasticsearch:/usr/share/elasticsearch/data
    networks:
      - igniterealtime_internal_network

  website:
    image: tootsuite/mastodon:v4.0.2
    env_file: 
      - application.env
      - database.env
    command: bash -c "bundle exec rails s -p 3000"
    restart: always    
    depends_on:
      - postgresql
#      - pgbouncer
      - redis
      - redis-volatile
      - elasticsearch
    ports:
      - '3001:3000'
    networks:
      - igniterealtime_internal_network
      - igniterealtime_external_network
    healthcheck:
      test: ['CMD-SHELL', 'wget -q --spider --proxy=off localhost:3000/health || exit 1']
    volumes:
      - uploads:/mastodon/public/system

  shell:
    image: tootsuite/mastodon:v4.0.2
    env_file: 
      - application.env
      - database.env
    command: /bin/bash 
    restart: "no"
    networks:
      - igniterealtime_internal_network
      - igniterealtime_external_network
    volumes:
      - uploads:/mastodon/public/system

  streaming:
    image: tootsuite/mastodon:v4.0.2
    env_file: 
      - application.env
      - database.env
    command: node ./streaming
    restart: always
    depends_on:
      - postgresql
#      - pgbouncer
      - redis
      - redis-volatile
      - elasticsearch
    ports:
      - '4001:4000'
    networks:
      - igniterealtime_internal_network
      - igniterealtime_external_network
    healthcheck:
      test: ['CMD-SHELL', 'wget -q --spider --proxy=off localhost:4000/api/v1/streaming/health || exit 1']

  sidekiq:
    image: tootsuite/mastodon:v4.0.2
    env_file: 
      - application.env
      - database.env
    command: bundle exec sidekiq
    restart: always
    depends_on:
      - postgresql
#      - pgbouncer
      - redis
      - redis-volatile
      - website
    networks:
      - igniterealtime_internal_network
      - igniterealtime_external_network
    healthcheck:
      test: ['CMD-SHELL', "ps aux | grep '[s]idekiq\ 6' || false"]
    volumes:
      - uploads:/mastodon/public/system

networks:
  igniterealtime_external_network:
  igniterealtime_internal_network:
    internal: true

volumes:
  postgresql:
    driver_opts:
      type: none
      device: /opt/mastodon_igniterealtime/database/postgresql
      o: bind    
  redis:
    driver_opts:
      type: none
      device: /opt/mastodon_igniterealtime/database/redis
      o: bind    
  elasticsearch:
    driver_opts:
      type: none
      device: /opt/mastodon_igniterealtime/database/elasticsearch
      o: bind    
  uploads:
    driver_opts:
      type: none
      device: /opt/mastodon_igniterealtime/web/system
      o: bind

**My nginx file :** 
cat mastodon_igniterealtime 
map $http_upgrade $connection_upgrade {
  default upgrade;
  ''      close;
}

upstream backend_igniterealtime {
    server 127.0.0.1:3001 fail_timeout=0;
}

upstream streaming_igniterealtime {
    server 127.0.0.1:4001 fail_timeout=0;
}

proxy_cache_path /var/cache/nginx_igniterealtime levels=1:2 keys_zone=CACHE_igniterealtime:10m inactive=7d max_size=1g;

#server {
#  listen 94.103.99.199:80;
#  server_name toot.igniterealtime.org;
#  location / { return 301 https://$host$request_uri; }
#}

server {
  listen 94.103.99.199:443 ssl http2;
  server_name toot.igniterealtime.org;

  ssl_protocols TLSv1.2 TLSv1.3;
  ssl_ciphers HIGH:!MEDIUM:!LOW:!aNULL:!NULL:!SHA;
  ssl_prefer_server_ciphers on;
  ssl_session_cache shared:SSL:10m;
  ssl_session_tickets off;

  ssl_certificate     /etc/letsencrypt/live/toot.igniterealtime.org/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/toot.igniterealtime.org/privkey.pem;

  keepalive_timeout    70;
  sendfile             on;
  client_max_body_size 80m;

  root /opt/mastodon/public;

  gzip on;
  gzip_disable "msie6";
  gzip_vary on;
  gzip_proxied any;
  gzip_comp_level 6;
  gzip_buffers 16 8k;
  gzip_http_version 1.1;
  gzip_types text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript image/svg+xml image/x-icon;

  add_header Strict-Transport-Security "max-age=31536000" always;

  location / {
    try_files $uri @proxy;
  }

  location ~ ^/(system/accounts/avatars|system/media_attachments/files) {
    add_header Cache-Control "public, max-age=31536000, immutable";
    add_header Strict-Transport-Security "max-age=31536000" always;
    root /opt/mastodon/;
    try_files $uri @proxy;
  }

  location ~ ^/(emoji|packs) {
    add_header Cache-Control "public, max-age=31536000, immutable";
    add_header Strict-Transport-Security "max-age=31536000" always;
    try_files $uri @proxy;
  }

  location /sw.js {
    add_header Cache-Control "public, max-age=0";
    add_header Strict-Transport-Security "max-age=31536000" always;
    try_files $uri @proxy;
  }

  location @proxy {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";
    proxy_pass_header Server;

    proxy_pass http://backend_igniterealtime;
    proxy_buffering on;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    proxy_cache CACHE;
    proxy_cache_valid 200 7d;
    proxy_cache_valid 410 24h;
    proxy_cache_use_stale error timeout updating http_500 http_502 http_503 http_504;
    add_header X-Cached $upstream_cache_status;
    add_header Strict-Transport-Security "max-age=31536000" always;

    tcp_nodelay on;
  }

  location /api/v1/streaming {
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Proxy "";

    proxy_pass http://streaming;
    proxy_buffering off;
    proxy_redirect off;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;

    tcp_nodelay on;
  }

  error_page 500 501 502 503 504 /500.html;
}

### Specifications

Mastodon 4.0.2 docker from pre-build 4.0.2 images

**Comment by DalinSeivewright** [comment]
I am also having this issue and am unable to resolve it.  Attempts to create a websocket result in a 401 Unauthorized in both Brave and Firefox and I've attempted several things (i.e. turning off Braves built-in protections) in an attempt to resolve it.

I am using the latest Docker 4.0.2 image from https://github.com/linuxserver/docker-mastodon which uses nginx with some slight differences with its nginx configuration but I am attempted to modify it to match https://github.com/mastodon/mastodon/blob/main/dist/nginx.conf with no success.

The web socket connection is specifically made against `my-instance.com/api/v1/streaming/?` and as I understand it, I should be expecting the access token to be appended to it as a query parameter, which I assume is causing this failure.

In addition, both Firefox and Chrome report a CSP violation due to `style-src` (unsafe-inline, hash or nonce).  Neither browser is very informative on what the exact content is in violation.  The CSP is probably caused due to the fact that I am using S3 for serving media and am using the S3_ALIAS_HOST to set the CDN Proxy for the bucket.  The CSP headers only include my base domain, and do not include media.my-instance.com.  Only mentioning this as it may be relevant as to why the access token is not accessible.

No errors appear in my server logs regarding any of this - seems to be purely a client-side issue.  The only other issues in the browser console are preload failure warnings because "request credentials mode does not match".  This mentions the crossorigin attribute so likely due to the S3_ALIAS_HOST as well.

Relevant config:
version=4.0.2
LOCAL_HOST=my-instance.com
S3_ALIAS_HOST=media.my-instance.com
ES_ENABLED=false

**Comment by shleeable** [comment]
@DalinSeivewright did you solve this?

**Comment by DalinSeivewright** [comment]
> @DalinSeivewright did you solve this?

I did not.  I attempted to tinker with it further but gave up.  I'm not OP though...  I just added some additional info in case these two issues were the same problem

**Comment by vmstan** [comment]
Are any of you using SSL back to your Postgres database?

**Comment by Rilgamon** [comment]
Migrated to a new server  and run into the same problem.  Hometimeline updates, but federated or others require a reload with this problem. Cant find helpful logentries ...

[weblog]
` - - [01/Jan/2023:16:01:06 +0100] "GET /api/v1/streaming/? HTTP/1.0" 404 21 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0"`

[syslog]
```
Jan  1 16:08:41 mastodon node[43840]: verb Subscribe timeline:access_token:102
Jan  1 16:08:41 mastodon node[43840]: verb Subscribe timeline:system:1
Jan  1 16:08:41 mastodon node[43840]: verb 657319a4-ee58-482f-a2be-ccf33e21bdb7 Subscription error: Unknown stream type
Jan  1 16:08:41 mastodon node[43840]: verb Unsubscribe timeline:access_token:102
Jan  1 16:08:41 mastodon node[43840]: verb Unsubscribe timeline:system:1
Jan  1 16:08:46 mastodon node[43840]: verb Subscribe timeline:access_token:102
Jan  1 16:08:46 mastodon node[43840]: verb Subscribe timeline:system:1
Jan  1 16:08:46 mastodon node[43840]: verb 2f1b7d45-9ea9-4d05-824d-01a2295332d1 Subscription error: Unknown stream type
Jan  1 16:08:46 mastodon node[43840]: verb Unsubscribe timeline:access_token:102
Jan  1 16:08:46 mastodon node[43840]: verb Unsubscribe timeline:system:1
Jan  1 16:08:49 mastodon node[43840]: verb Subscribe timeline:access_token:102
Jan  1 16:08:49 mastodon node[43840]: verb Subscribe timeline:system:1
Jan  1 16:08:49 mastodon node[43840]: verb e039b4cb-1b36-46ee-b9e6-6456564598bb Subscription error: Unknown stream type
Jan  1 16:08:49 mastodon node[43840]: verb Unsubscribe timeline:access_token:102
Jan  1 16:08:49 mastodon node[43840]: verb Unsubscribe timeline:system:1

```

**Comment by Rilgamon** [comment]
I found that for me this was fixed when I did not use a reverse proxy. And the connection gets a 101-answer.
Copying the headers from this to the api/v1/streaming block in nginx allowed the proxy version to work,too.

```
    proxy_set_header Upgrade "websocket";
    proxy_set_header Connection "upgrade";
```

**Reply by sbug0** [reply]
THANK YOU SO MUCH! 🤗🙌🏻

**Comment by ClaudeStabile** [comment]
Hello,

I finally fixed this issue with several actions :+1: 
1-.upgrade all mastodon components rework my nginx site file including latest changes from mastodon template and my amendements to support several mastodon instance on same host 
2-Add in application.env file this line : 
ALLOWED_HOSTS=toot.igniterealtime.org,94.103.99.199   # IP used on  my server that resolved to 94.103.99.199

And upgrade all  involve system components docker, nginx 
upgrade mstodon to latest current for me is 4.3.8 
I do hope it can helps some of you here facing similar issues
![screenshot36](https://github.com/user-attachments/assets/d4af5178-872e-449d-8618-27e86c5d4378)


All this together seemed to have fixed all issues i had including this one
